models/base_test.py

```diff
@@ -0,0 +1,139 @@
+            print(line, end="")
+            limit += 1
+    print("\n------------------------------------------------------------\n")
+
+
+class ModelTest(unittest.TestCase):
+    """Base class for testing Certifai Prediction Service Examples. Each example will typically include multiple
+    scenarios where:
+
+        1) a flask server is launched as a background process (via the Certifai Model SDK)
+        2) a Certifai Scan is launched (or just a plain Python script) is launched in the foreground that calls (1)
+
+    Each process that is launched in the foreground, (2), is expected to complete with a 0 exit code. Each process
+    launched in the background (1) are expected to be run until explicitly killed.
+
+    The following functions should cover scenarios that run plain Python Scripts::
+
+        run_standalone_python_script(python_script)
+        run_python_app_test(model_app, python_script)
+
+    The following functions should cover scenarios that involve running a Certifai Scan::
+
+        run_model_and_definition_test('app_dtree.py', 'my-definition.yaml')
+        run_model_and_scan('app_dtree.py', 'my-definition.yaml')
+        run_model_and_explain('app_dtree.py', 'my-definition.yaml', fast=True)
+    """
+    SLEEP_TIME = 5        # 5 seconds
+    TERMINATION_TIME = 5  # 5 seconds
+    DEFAULT_TEST_TIMEOUT = 2 * 60       # 2 minutes
+    DEFAULT_SCAN_TIMEOUT = 60 * 60 * 2  # 2 hours
+    bg = None
+
+    def _run_in_foreground(self, command: Sequence[str], timeout: Optional[int] = None):
+        try:
+            # Run process and wait until it completes
+            process = subprocess.run(command, shell=False, capture_output=True, timeout=timeout, text=True)
+            process.check_returncode()
+        except subprocess.TimeoutExpired as te:
+            error = f"\nProcess did not finish within expected time (command={te.cmd}, timeout={te.timeout} seconds). Error: {str(te)}"
+            capture_err_and_out(te.stderr, te.stdout)
+            self.fail(error)
+        except subprocess.CalledProcessError as ce:
+            error = f"\nProcess finished with non-zero exit code (command={ce.cmd}, code={ce.returncode}). Error: {str(ce)}"
+            capture_err_and_out(ce.stderr, ce.stdout)
+            self.fail(error)
+
+    @contextlib.contextmanager
+    def _run_in_background(self, command: Sequence[str]):
+        with tempfile.NamedTemporaryFile(mode='w+') as stdout, tempfile.NamedTemporaryFile(mode='w+') as stderr:
+            try:
+                p = subprocess.Popen(command, shell=False, stdout=stdout, stderr=stderr, stdin=subprocess.DEVNULL,
+                                     close_fds=True, text=True)
+                yield
+            except Exception:
+                # WARNING: Killing the subprocess may not kill any workers spawned by the process (e.g. gunicorn!)
+                p.kill()
+                p.wait()
+                capture_output(stdout.name, stderr.name)
+                raise
```

Comment:
The shutdown functionality for the prediction services is kinda wonky right now, especially when testing. It seemed like the best way to handle it was by using Python's subprocess module & just killing the parent process (and for the sake of our examples, the apps run with the dev server so we never risk leaving zombie workers). 

* The builtin `/shutdown` endpoint only works for `gunicorn` servers (when `production=True`) - because a shell script (in the package) will kill all the processes.
* Not all tests will call the `/shutdown` endpoint - like when running python scans. Nor do we expect all the servers to be run with `production=True`
* Python's `kill()` here will kill the process, but if you are gunicorn, then the spawned workers are left over